micro-rdk/src/esp32/camera.rs

```diff
@@ -1,117 +1,254 @@
+                frame_size,
+                jpeg_quality,
+                // Number of frame buffers to be allocated.
+                // If more than one, then each frame will be acquired (double speed)
+                // when pixel_format == jpeg, fb_count > 1 goes to continuous mode, may need to adjust
+                // xclk_freq_hz down to 10_000_000.
                 fb_count: 1,
                 grab_mode: 0,
                 fb_location: 0,
-                sccb_i2c_port: 0,
+                sccb_i2c_port,
             },
-            last_grab: t.now(),
-        }
+        };
+
+        esp!(unsafe { esp_camera_init(&cam.config) })
+            .map_err(|e| CameraError::InitError(Box::new(e)))?;
+        Ok(Arc::new(Mutex::new(cam)))
     }
-    pub fn setup(&self) -> Result<(), CameraError> {
-        let ret = (unsafe { crate::esp32::esp_idf_svc::sys::esp_camera_init(&self.config) })
-            as crate::esp32::esp_idf_svc::sys::esp_err_t;
-        let ret = crate::esp32::esp_idf_svc::sys::EspError::convert(ret);
-        ret.map_err(|e| CameraError::CameraInitError(e.into()))
+
+    fn get_frame(&mut self) -> Result<Esp32CameraFrameBuffer, CameraError> {
+        Esp32CameraFrameBuffer::get().ok_or(CameraError::FailedToGetImage)
     }
-    pub fn get_cam_frame(&self) -> Option<*mut crate::esp32::esp_idf_svc::sys::camera_fb_t> {
-        let ptr = (unsafe { crate::esp32::esp_idf_svc::sys::esp_camera_fb_get() })
-            as *mut crate::esp32::esp_idf_svc::sys::camera_fb_t;
-        if ptr.is_null() {
-            None
-        } else {
-            Some(ptr)
+}
+
+impl Camera for Esp32Camera {
+    fn get_image(&mut self, mut buffer: BytesMut) -> Result<BytesMut, CameraError> {
+        let frame = self.get_frame()?;
+        if frame.len() > buffer.capacity() {
+            return Err(CameraError::ImageTooBig);
         }
+        let msg = camera::v1::GetImageResponse {
+            mime_type: "image/jpeg".to_string(),
+            image: frame.as_bytes(),
+        };
+        // safety: message must be encoded before the frame is dropped from scope
+        msg.encode(&mut buffer).unwrap();
+
+        return Ok(buffer);
     }
-    pub fn return_cam_frame(
-        &self,
-        frame: Option<*mut crate::esp32::esp_idf_svc::sys::camera_fb_t>,
-    ) {
-        if let Some(ptr) = frame {
-            unsafe { crate::esp32::esp_idf_svc::sys::esp_camera_fb_return(ptr) }
+
+    fn render_frame(&mut self, mut buffer: BytesMut) -> Result<BytesMut, CameraError> {
+        let frame = self.get_frame()?;
+        if frame.len() > buffer.capacity() {
+            return Err(CameraError::ImageTooBig);
         }
+        let msg = HttpBody {
+            content_type: "image/jpeg".to_string(),
+            data: frame.as_bytes().to_vec(),
+            ..Default::default()
+        };
+        // safety: message must be encoded before the frame is dropped from scope
+        msg.encode(&mut buffer).unwrap();
```

Comment:
It isn't a problem with your code, but it does feel like the camera data gets copied a lot here trying to get it into the proto message. I'm not sure you can do better though. Same for the `get_image` call above.

Reply:
I know what you mean. I need to run more tests later to determine how many times data is being (re)written and everything. Even the `to_vec()` in L186 feels suspect.  

Ideally, the message can be encoded in-place, where it has already been written to by the camera so we can send it directly.

Reply:
Other than the `to_vec` call, I don't understand where the additional copying is coming from. The `as_bytes` function consumes the frame and the `Bytes` struct is really just an iterator. Is it in the call to `encode`?

Reply:
Yes, the call to encode.

Reply:
I'm not sure how we can prevent this though and form the message buffer in-place as the actual buffer containing the bytes is managed by `esp-cam` via DMA. When we encode, the bytes are then copied over from the camera's frame buffer into our webrtc/grpc buffer.

Reply:
OK, thanks for looking into it.